### ISSUE TITLE
Improve definition of `@sameas`

### ISSUE THREAD

**Issue by pe-ro** [music-encoding/music-encoding#1056]
`@sameas` is intended to allow **_a single markup entity_** (aka element) to exist in multiple places in the markup at the same time.

The documentation for `@sameas` says: "Points to an element that is the same as the current element but is not a literal copy of the current element."  What's unsaid here is that in order for an element to be the same as another, at the very least the elements must share the same name.  

Here's an example of the intended usage --

```
In layer 1:
<rest dur="4" tstamp="4" xml:id="rest_2202"/>
...
In layer 2:
<rest xml:id="rest_2223" sameas="#rest_2202" />
```

`@sameas` on the 2nd `<rest>` indicates that it is the same rest as the one referred to but the markup requires double encoding.

Here's an example of incorrect usage --

```
<composer>
  <persName auth="GND" auth.uri="https://d-nb.info/gnd/118584596" codedval="118584596"
    label="MI-PID_39409" xml:id="WAM">Wolfgang Amadé Mozart</persName>
</composer>
...
<composer sameas="#WAM" />
```

The intention seems to be indicate that the composer in the 2nd case is the same as the one in the 1st case.  There are a couple of problems here, though.  First, the elements bearing the id "WAM" and the one bearing the reference "#WAM" have different names. (Note: this issue could be "fixed" by putting the `@xml:id="WAM"` on the first `<composer>` element.)

However, there's a second, more thorny problem.  That is, the **_elements_** bearing `@xml:id` and `@sameas` are **_not  the same markup entity_** the way the the `<rest>` elements are in the example above.

Currently, the better way to indicate that the composer (Mozart) is the **_same person_** in both cases (or at least that the personal names for that person are the same) would be to use the same `@auth.uri` value, e.g.,

```
<composer>
  <persName auth="GND" auth.uri="https://d-nb.info/gnd/118584596" codedval="118584596"
    label="MI-PID_39409" xml:id="WAM">Wolfgang Amadé Mozart</persName>
</composer>
...
<composer>
  <persName auth.uri="https://d-nb.info/gnd/118584596" />
</composer>
```

Currently, only certain elements allow the `att.authorized` attribute class so it's not possible to use `@auth.uri` on `<composer>` .  It may be useful to discuss making `att.authorized` available in more places, but that's probably better addressed by a separate issue/bug report.

In the meantime, to encourage correct usage, I suggest changing the documentation for `@sameas` to "Points to an element that is the same markup entity as the current element. The referring element and the element referred to don't have to share all attributes."

Also, a Schematron rule should be implemented to force the element names of the referring element and the element referred to to be the same.

**Comment by bwbohl** [comment]
Great catch @pe-ro!
I think it would also be useful to give an example of what would happen with shared or unshared attributes, e.g. `@xml:id`could never be shared / the same, what would be the meaning if the pointing element would share an attribute with the element pointed to, or what if it would have additional attributes. Should this even be allowed, or should we add that the pointing element may not have other attributes other than `@sameas` and `@xml:id` (and maybe some more?). This could be validated with a Schematron rule, too.

**Comment by frakel** [comment]
Thank you @pe-ro for commenting on this! A nice example of how errors promote the development of MEI 😉. The current implementation of `@sameas` is definitely error prone and needs at least to have some schematron rule in this context in a future MEI release. 
I would suggest to just allow `@xml:id` besides of `@sameas`. Otherwise the meaning of @sameas is weakened or obfuscated. Furthermore it is possible to point to same entities also via other attributes as @pe-ro mentioned (e.g. `@auth-uri`) where attributes could be mixed...

**Comment by ahankinson** [comment]
I think the existing documentation for `@sameas` is good, but just needs further qualification? Introducing the term "markup entity" seems to obscure the fact that they should simply be the same element name. Perhaps:

"Points to an element that is the same as the current element but is not a literal copy of the current element. The referring element and the element referred to do not have to share all attributes, but they must share the same element name."

**Comment by ahankinson** [comment]
@bwbohl @frakel we probably also need to allow different positioning attributes on elements with `@sameas` since a rest, for example, can be "the same" as another rest, but with a different layout.

**Comment by lpugin** [comment]
> [...] can be "the same" as another rest, but with a different layout.

Then it is not the same... I think that is a case where you would use `@copyof`.

**Comment by ahankinson** [comment]
If you have a score and parts, and want to point a rest from the score to its location in the part, then it's arguably "the same" rest at a different location.

Having no different attributes is still possible; if you're restricting your definition of the same element at the same location at the same time, then they would completely overlap and wouldn't be much use.

**Comment by lpugin** [comment]
One example where I think it is useful to use `@sameas` for specifying the it is the same element at the same location at the same time can be when you have a clef change that is appearing in two layers but is actually appearing only once. They are both literally the same:
![image](https://user-images.githubusercontent.com/689412/205889858-20be907d-1d2c-4cd3-b22b-e5c5465351f6.png)
```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/4.0.0/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="4.0.0">
   <meiHead>
      <fileDesc>
         <titleStmt>
            <title>Example use of @sameas on note and clef</title>
         </titleStmt>
         <pubStmt>
            <date>2017-05-15</date>
         </pubStmt>
         <notesStmt>
            <annot>Note and clef can refer to another element using @sameas.</annot>
         </notesStmt>
      </fileDesc>
      <encodingDesc>
         <appInfo>
            <application version="unknown">
               <name>Verovio</name>
            </application>
         </appInfo>
      </encodingDesc>
   </meiHead>
   <music>
      <body>
         <mdiv>
            <score>
               <scoreDef midi.bpm="120">
                  <staffGrp bar.thru="true">
                     <staffDef n="1" lines="5" clef.shape="G" clef.line="2" key.sig="2s" />
                  </staffGrp>
               </scoreDef>
               <section>
                  <measure>
                     <staff n="1">
                        <layer n="1">
                           <note xml:id="n1" dur="4" oct="6" pname="c" />
                           <clef xml:id="c1" shape="F" line="4" />
                           <note dur="4" oct="3" pname="c" />
                        </layer>
                        <layer n="2">
                           <note sameas="#n1" />
                           <clef sameas="#c1" />
                           <note dur="4" oct="3" pname="c" />
                        </layer>
                     </staff>
                  </measure>
               </section>
            </score>
         </mdiv>
      </body>
   </music>
</mei>
```
Another case is when two instruments "share" the same notes. You can have the two layers encoded, and the notes in the second layer are all the same as the ones in the other layer because they do appear only once in the score.

Wouldn't be the case you describe more a case where you would use `@corresp`? I think that would be more correct than saying they are the same.

**Comment by o-sapov** [comment]
What about shared / unshared descendant nodes?
Consider the case when another form of the composer name would be used. Is it still appropriate to use `@sameas` here?

```xml
     <composer>
          <persName auth="GND" auth.uri="https://d-nb.info/gnd/118584596" codedval="118584596" label="MI-PID_39409" xml:id="WAM">Wolfgang Amadé Mozart</persName>
        </composer>
```



```xml
<composer>
              <persName sameas="#WAM">Wolfgango Amedeo Mozart</persName>
            </composer>
```

Note that `@auth.uri="https://d-nb.info/gnd/118584596"` points to GND where several name forms are listed.

**Comment by o-sapov** [comment]
Otherwise, it would be acceptable to point to the highest semantic level expecting the descendant nodes to be the  _same_. Consider the example when _Clarini_ share a measure rest, meaning that layers 1 and 2 are same.

<img width="300" alt="nma_k310-001_m68" src="https://user-images.githubusercontent.com/24490898/205938292-405276d8-cfb3-491a-bdac-24a531cc2df7.png">

<img width="300" alt="nma_k310-001_m68" src="https://user-images.githubusercontent.com/24490898/205938531-e394f580-7b16-4ca5-bb86-bbd566c00594.png">

**Comment by ahankinson** [comment]
@o-sapov That also seems like a good example of two things that can have different attributes but be considered the same -- `n="1"` and `n="2"` can be different targets for `<layerDef>` definitions, even though they are the "same."

**Comment by pe-ro** [comment]
Hi @o-sapov,

`@sameas` on `<persName>` is contradictory -- the two names are definitely not the same.

I think `@corresp` is a better choice to indicate a relationship between the two names, i.e.,
```
<composer>
  <persName auth="GND" auth.uri="https://d-nb.info/gnd/118584596" label="MI-PID_39409" xml:id="WAM">Wolfgang Amadé Mozart</persName>
</composer>
...
<composer>
  <persName corresp="#WAM">Wolfgango Amedeo Mozart</persName>
</composer>
```

The following would also be an acceptable short hand, but creates a relationship between the two composer elements --

```
<composer xml:id="WAM">
  <persName auth="GND" auth.uri="https://d-nb.info/gnd/118584596" codedval="118584596" label="MI-PID_39409">Wolfgang Amadé Mozart</persName>
</composer>
...
<composer corresp="#WAM">
  <persName>Wolfgango Amedeo Mozart</persName>
</composer>
```

However, to be absolutely clear, the best option is create a relationship between the different names and an authority record --

```
<composer>
  <persName auth="GND" auth.uri="https://d-nb.info/gnd/118584596" label="MI-PID_39409" xml:id="WAM">Wolfgang Amadé Mozart</persName>
</composer>
...
<composer>
  <persName auth="GND" auth.uri="https://d-nb.info/gnd/118584596">Wolfgango Amedeo Mozart</persName>
</composer>
```

By the way, it's not necessary to use `@codedval`, as it adds nothing that isn't already provided in the  value of `@auth.uri`.

**Comment by pe-ro** [comment]
In my opinion, when `@sameas` is used, it's the elements that are the same thing.  Some, but probably not all, attributes may vary depending on context.
